e2e: cannot delete directory - sometimes / race condition?

**Version of Singularity**

main

**Describe the bug**

e2e tests in CircleCI sometimes fail with...

```
=== NAME  TestE2E
    fileutil.go:51: failed to delete temporary directory: fstatat /tmp/stest.1651457626/e2e-imgcache-2702428687/rootfs-3909637960/root: permission denied
    suite.go:134: Test failed, not removing /tmp/stest.1651457626
--- FAIL: TestE2E (555.53s)
```

The path indicates this is related to a FUSE mounted image.

It has been impossible to reproduce this outside of CircleCI so far. I have tried:

- Low vCPU counts (1 ->4)
- High physical CPU counts (16)
- Variety of distro (RHEL 7, 8, 9. SLES 12, 15. Ubuntu 20.04, 22.04)

In the past, similar issues have generally indicated a race condition in the e2e-test. It's either that, or we have a path where the FUSE cleanup isn't firing / is getting killed in an unexpected way.

Not letting this block the RC given I cannot reproduce outside of e2e-tests and it's going to be difficult to track down.

--- a/internal/app/starter/host.go
+++ b/internal/app/starter/host.go
@@ -72,7 +72,8 @@ func CleanupHost(cleanupSocket int, e *engine.Engine) {
 		_, err := conn.Read(data)
 		// Clean shutdown - master should be notified after cleanup.
 		if err == nil {
-			tc <- syscall.SIGHUP
+			tc <- syscall.SIGUSR1
+			return
 		}
 		// Unclean shutdown - master not available to notify after cleanup.
 		tc <- syscall.SIGTERM
@@ -88,7 +89,7 @@ func CleanupHost(cleanupSocket int, e *engine.Engine) {
 	err = e.CleanupHost(context.TODO())
 
 	// If we are in a clean (master initiated) shutdown, notify master of any cleanup failure.
-	if err != nil && sig == syscall.SIGHUP {
+	if err != nil && sig == syscall.SIGUSR1 {
 		if _, err := conn.Write([]byte{'f'}); err != nil {
 			sylog.Debugf("Could not write to master: %s", err)
 		}
@@ -101,7 +102,7 @@ func CleanupHost(cleanupSocket int, e *engine.Engine) {
 	}
 
 	// If we are in a clean (master initiated) shutdown, notify master of the cleanup success.
-	if sig == syscall.SIGHUP {
+	if sig == syscall.SIGUSR1 {
 		if _, err := conn.Write([]byte{'c'}); err != nil {
 			sylog.Debugf("Could not write to master: %s", err)
 			sylog.Debugf("Exiting CleanupHost - Master socket write failure")

--- a/internal/app/starter/stage_linux.go
+++ b/internal/app/starter/stage_linux.go
@@ -6,6 +6,7 @@
 package starter
 
 import (
+	"context"
 	"net"
 	"os"
 
@@ -25,10 +26,16 @@ func StageOne(sconfig *starterConfig.Config, e *engine.Engine) {
 	sylog.Debugf("Entering stage 1\n")
 
 	if err := e.PrepareConfig(sconfig); err != nil {
+		if cleanErr := e.CleanupHost(context.TODO()); cleanErr != nil {
+			sylog.Errorf("While running host cleanup tasks: %s", cleanErr)
+		}
 		sylog.Fatalf("%s\n", err)
 	}
 
 	if err := sconfig.Write(e.Common); err != nil {
+		if cleanErr := e.CleanupHost(context.TODO()); cleanErr != nil {
+			sylog.Errorf("While running host cleanup tasks: %s", cleanErr)
+		}
 		sylog.Fatalf("%s", err)
 	}
 

--- a/internal/pkg/runtime/engine/engine_linux.go
+++ b/internal/pkg/runtime/engine/engine_linux.go
@@ -88,20 +88,24 @@ type Operations interface {
 	// a hybrid workflow (e.g. fakeroot), then there is no privileged saved uid
 	// and thus no additional privileges can be gained.
 	CleanupContainer(context.Context, error, syscall.WaitStatus) error
-	// PostStartHost is called after the container process has been executed. It is
-	// run in the process forked from starter before namespace setup etc. and
-	// will perform any cleanup in the host mount namespace at time of CLI
-	// execution.
+	// PostStartHost is called after the container process has been executed. It
+	// is run in the POST_START_HOST process forked from starter before
+	// namespace setup etc. and will perform any cleanup in the host mount
+	// namespace at time of CLI execution.
 	//
-	// No additional privileges can be gained during this call, as privileges
-	// are dropped permanently after forking in starter.
+	// No additional privileges can be gained during this call in the setuid
+	// flow as privileges are dropped permanently after forking in starter.
 	PostStartHost(context.Context) error
-	// CleanupHost is called after CleanupContainer, before master exits.
-	// It is run in the process forked from starter before namespace setup etc. and
-	// will perform any cleanup in the host mount namespace at time of CLI execution.
+	// CleanupHost is called on container exit or startup failure to perform any
+	// required cleanup in the host mount namespace at time of CLI execution.
 	//
-	// No additional privileges can be gained during this call, as privileges are
-	// dropped permanently after forking in starter.
+	// If container creation fails early, in STAGE 1, it will be called directly
+	// from STAGE 1. Otherwise it is run in the CLEANUP_HOST PROCESS, triggered
+	// by master, or the SIGKILL parent death signal.
+	//
+	// No additional privileges can be gained during this call in the setuid
+	// flow, as privileges are dropped permanently in both STAGE1 and
+	// CLEANUP_HOST after forking in starter.
 	CleanupHost(context.Context) error
 }
 

--- a/internal/pkg/runtime/engine/singularity/host_linux.go
+++ b/internal/pkg/runtime/engine/singularity/host_linux.go
@@ -26,8 +26,9 @@ func (e *EngineOperations) PostStartHost(ctx context.Context) (err error) {
 }
 
 // CleanupHost cleans up a SIF FUSE image mount and the temporary directory that
-// holds it, if it exists after container exit. It is called from a HOST_CLEANUP process that exists
-// in the original host namespaces.
+// holds it. If container creation fails early, in STAGE 1, it will be called
+// directly from STAGE 1. Otherwise, it will be called from a CLEANUP_HOST
+// process, when the container cleanly exits, or is killed.
 func (e *EngineOperations) CleanupHost(ctx context.Context) (err error) {
 	tmpDir := e.EngineConfig.GetDeleteTempDir()
 	if e.EngineConfig.GetImageFuse() && tmpDir != "" {

